### PR TITLE
Create pipeline to handle image promotion from private registry to public

### PR DESCRIPTION
**Pull request by sjonpaulbrown**
## Description
The following PR creates a new workflow that will centralize the process for building and releasing Flow docker images.

## Example Runs
* [Public Build](https://github.com/onflow/flow-go/actions/runs/14605211390)
* [Secure Build](https://github.com/onflow/flow-go/actions/runs/14605685087)

## 🔧 GitHub Actions Workflow: Build & Promote Docker Images to Public Registry

This workflow is designed to **build, tag, and publish Docker images** for the `flow-go` project using GitHub Actions. It supports **both public and secure builds**, and provides a mechanism to **promote successfully built images from a private registry to a public registry**.

---

### 🎯 **Goals & Intentions**

- **Build and push Docker images** for various `flow-go` roles (e.g., access, collection, consensus, etc.) based on a specified Git tag.
- **Support two build modes**:
  - **Public Build** (`secure-build: false`): Pulls from the public flow-go repository, builds images using public dependencies, and pushes to a private registry.
  - **Secure Build** (`secure-build: true`): Executes secure, cross-repo builds that may depend on private components. These builds are orchestrated through a triggered workflow in a private repo.
- **Promote built images** from the private Google Artifact Registry to a **public container registry** once builds are successful.
- **Require Approvals** from the `flow-protocol-release` team to promote an image from the private registry to the public

---

### 🚀 **Workflow Structure**

1. **Trigger Input Parameters**
   - `secure-build`: Determines if the build should use private dependencies and cross-repo logic.
   - `tag`: Git tag to check out and use for versioning all generated Docker images.

2. **Jobs Overview**
   - **`public-build`**:
     - Runs if `secure-build` is `false`.
     - Uses a matrix of `make` commands to build and push Docker images for various roles.
     - Builds run in parallel for efficiency.
   - **`secure-build`**:
     - Runs if `secure-build` is `true`.
     - Triggers a separate workflow in a private GitHub repo to handle secure builds for each role.
     - Uses matrix strategy to parallelize secure builds per role.
   - **`promote-images`**:
     - Depends on the successful completion of either the `public-build` or `secure-build` jobs.
     - Promotes images for each role independently to a public registry.
     - Utilizes a reusable promotion action to transfer images from the private to public registry under specific tags.
     - Each role-specific promotion job is executed **independently**, allowing granular control over which components are released publicly.
     - The **`flow-protocol-release` team** is responsible for **approving and overseeing** the promotion jobs. These jobs are gated by environment-based approval workflows to ensure controlled and intentional releases.

---

### 🛡️ **Security & Permissions**

- Uses **Google Cloud authentication** to securely push and promote images via the Artifact Registry.
- Sensitive secrets such as service account credentials and GitHub tokens are sourced from repository secrets.
- Each job is scoped to a relevant environment for additional control and auditing.
- Promotions to the public registry are subject to **manual approval**, providing the `flow-protocol-release` team full authority over what is released.

---

### 🧪 **Benefits**

- Ensures consistent Docker image generation across roles.
- Minimizes latency via parallelized builds.
- Enables gated secure builds for critical images.
- Promotes image visibility by syncing with a public registry.
- Supports **role-specific promotion**, enabling selective releases of flow node images

---

### 📝 Notes
- Secure builds require pre-configured cross-repo workflows and the appropriate permissions to trigger them.
- Promotion environments should be reviewed and approved by the `flow-protocol-release` team before final public release.

### Documentation
- [Technical Details for Workflow](https://www.notion.so/flowfoundation/Technical-Details-for-flow-go-images-1611aee1232480e396c4fa9906e3aa61)
- [User instructions for Workflow](https://www.notion.so/flowfoundation/How-to-Build-Images-in-flow-go-1611aee12324809cb45bede927f7b279)

### CODE REVIEW

**Review by peterargue — APPROVED** [review]

File: .github/workflows/image_builds.yml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    name: Execute public repo build & push to private artifact registry 
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        # We specify all of the potential build commands for each role. 
+        # This allows us to build and push all images in parallel, reducing the overall build time.
+        # The matrix is defined to include all roles & image types that we want to build and push.
+        # These commands are targets defined in the Makefile of the repository.
+        build_command:
```
Comment (Kay-Zee) [inline .github/workflows/image_builds.yml:37]: wow we have a lot of these now

File: .github/workflows/image_builds.yml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        with:
+          client_payload: '{"role": "${{ matrix.role }}", "tag": "${{ inputs.tag }}"}'
+          github_token: ${{ secrets.SECURE_BUILDS_TOKEN }}
+          github_user: ${{ secrets.SECURE_BUILDS_TOKEN_USER }}  
+          owner: 'onflow'
+          repo: ${{ secrets.SECURE_BUILDS_REPO }}  
+          ref: master-private 
+          workflow_file_name: 'secure_build.yml'
+
+  promote-images:
```
Comment (Kay-Zee) [inline .github/workflows/image_builds.yml:128]: are we combining these initially just to mimic the current process, and then splitting these eventually?

**Review by Kay-Zee — APPROVED** [review]

File: .github/workflows/image_builds.yml (hunk 2 above)
Comment (sjonpaulbrown) [reply .github/workflows/image_builds.yml:128]: Since the docker build process takes longer to execute, I decided to split up the build and execute them all in parallel. This drastically reduces the time to execute all of the builds. It takes it from roughly 12 mins to build all of the images to 3 mins.

For the promotion, the assumption is that we don't want to promote specific images for a node type. The promotion is by node type rather than by the individual images that were built. This ensures that when we promote a build that all of the images for that node type are pushed to the public registry. To do this, we pass in the tags that would have been built, and we download & push each tag.

Comment (sjonpaulbrown) [reply .github/workflows/image_builds.yml:128]: > Right now with this setup, would it be possible to build and push to the private registry without promoting?

Yes, this is possible.

With this setup, there are no reviews required for building & pushing to the private registry. The `secure-build` and `public-build` jobs handle building the images & pushing to the private registry. After either of the build jobs run, the `promotion` jobs wait for individual approvals before running. Each `<node_type> image promotion to public registry` environment requires an approval from the `flow-protocol-release` group. This blocks the promotion job from being run, and while ensuring that we can run each promotion job individually, the creation of individual environments enables separate owners to eventually approve the different builds. 

In summary, this enables anyone with access to this workflow to build & push images to the private registry, but promotion is blocked until a reviewer manually approves the promotion of the image.
